File: Java/src/Gray/Gray.java
package Gray;

import java.util.Arrays;
import java.util.Scanner;

public class Gray {
    public static int length, number;

    public static void main(String[] args) {
        char[][] gray = init();
        encode(gray);
        System.out.println("Number\tGray\tBinary");
        for (int i = 0; i < number; i++) {
            System.out.println(i + "\t\t" + new String(gray[i]) + "\t" + new String(decimalToBinary(i)));
        }
    }

    public static char[][] init() {
        System.out.print("Please enter the length of the gray code you want:");
        Scanner sc = new Scanner(System.in);
        length = sc.nextInt();
        number = (int) Math.pow(2, length);
        char[][] gray = new char[number][length];
        for (int i = 0; i < number; i++) {
            Arrays.fill(gray[i], '0');
        }
        return gray;
    }

    public static void encode(char[][] gray) {
        for (int i = 1; i <= length; i++) {
            singleOperate(i, gray);
        }
    }

    public static void singleOperate(int digital_number, char[][] gray) {
        if (digital_number == 1) {
            gray[1][length - 1] = '1';
        } else {
            int range = (int) Math.pow(2, digital_number);
            for (int i = 0; i < range / 2; i++) {
                for (int j = length - 1; j >= length - digital_number; j--) {
                    gray[range - 1 - i][j] = gray[i][j];
                    gray[range - 1 - i][length - digital_number] = '1';
                }
            }
        }
    }

    public static char[] decimalToBinary(int decimal) {
        char[] binary = new char[length];
        Arrays.fill(binary, '0');
        int index = length - 1;
        while (decimal > 1) {
            int quotient = decimal / 2;
            binary[index--] = decimal - quotient * 2 == 1 ? '1' : '0';
            decimal = quotient;
        }
        binary[index] = decimal == 1 ? '1' : '0';
        return binary;
    }
}

File: Python/Gray/Gray.py
length, number = 0, 0

def decimalToBinary(decimal):
    binary = ['0' for _ in range(length)]
    index = length - 1
    while decimal > 1:
        quotient = int(decimal / 2)
        binary[index] = '1' if decimal - quotient * 2 == 1 else '0'
        index -= 1
        decimal = quotient
    binary[index] = '1' if decimal == 1 else '0'
    return binary

def singleOperate(digital_number, gray):
    if digital_number == 1:
        gray[1][length - 1] = '1'
    else:
        r = 2 ** digital_number
        for i in range(int(r / 2)):
            for j in range(length - 1, length - digital_number - 1, -1):
                gray[r - 1 -i][j] = gray[i][j]
                gray[r - 1- i][length - digital_number] = '1'

def encode(gray):
    for i in range(1, length + 1, 1):
        singleOperate(i, gray)

def init():
    global length, number
    length = int(input("Please enter the length of the gray code you want:"))
    number = 2 ** length
    return [['0' for _ in range(length)] for _ in range(number)]

def main():
    gray = init()
    encode(gray)
    print("Number\tGray\tBinary")
    for i in range(number):
        print(f"{i}\t{''.join(gray[i])}\t{''.join(decimalToBinary(i))}")

if __name__ == "__main__":
    main()